File: medium/380. Insert Delete GetRandom O(1)/380. Insert Delete GetRandom O(1).py
import random


class RandomizedSet:

    def __init__(self):
        self.lst: list[int] = []
        self.obj: dict[int, int] = {}

    def insert(self, val: int) -> bool:
        if val in self.obj:
            return False

        self.lst.append(val)
        self.obj[val] = len(self.lst) - 1
        return True

    def remove(self, val: int) -> bool:
        if val in self.obj:
            idx_to_remove = self.obj[val]
            lastElement = self.lst[-1]

            self.lst[idx_to_remove] = lastElement
            self.obj[lastElement] = idx_to_remove

            self.lst.pop()
            del self.obj[val]
            return True

        return False

    def getRandom(self) -> int:
        return random.choice(list(self.obj))


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: medium/380. Insert Delete GetRandom O(1)/380. Insert Delete GetRandom O(1).ts
class RandomizedSet {
    lst: number[]
    indices: { [key in number]: number }

    constructor() {
        this.lst = []
        this.indices = {}
    }

    insert(val: number): boolean {
        if (val in this.indices)
            return false

        const newSize = this.lst.push(val)
        this.indices[val] = newSize - 1
        return true
    }

    remove(val: number): boolean {
        if (!(val in this.indices))
            return false

        const idxToRemove = this.indices[val]
        const lastElement = this.lst[this.lst.length - 1]

        this.lst[idxToRemove] = lastElement
        this.indices[lastElement] = idxToRemove

        this.lst.pop()
        delete this.indices[val]
        return true
    }

    getRandom(): number {
        return this.lst[Math.floor(Math.random() * this.lst.length)]
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */